File: Design-Patterns-Using-Kotlin/Adapter/AdapterPattern.kt
interface DesiredInterface {
    fun operation()
}

// Adapter class
class Adapter : DesiredInterface {
    private val adaptee = Adaptee()

    override fun operation() {
        adaptee.someOperation()
    }
}

// Adaptee class
class Adaptee {
    fun someOperation() {
        println("Adaptee someOperation() function called.")
    }
}

// Client Code
fun main() {
    val adapter: DesiredInterface = Adapter()
    adapter.operation()
}

/*
Adaptee someOperation() function called.
*/

File: Design-Patterns-Using-Go/Adapter/AdapterPattern.go
package main

import "fmt"

// DesiredInterface interface
type DesiredInterface interface {
	Operation()
}

// Adapter struct
type Adapter struct {
	adaptee *Adaptee
}

func (a *Adapter) Operation() {
	a.adaptee.SomeOperation()
}

// Adaptee struct
type Adaptee struct{}

func (a *Adaptee) SomeOperation() {
	fmt.Println("Adaptee SomeOperation() function called.")
}

// Client Code
func main() {
	adapter := Adapter{adaptee: &Adaptee{}}
	adapter.Operation()
}

/*
Adaptee SomeOperation() function called.
*/